### ISSUE TITLE
TypedArray access shouldn't walk prototype chain

### ISSUE THREAD

**Issue by fatcerberus** [svaarala/duktape#257]
I just stumbled across this SpiderMonkey bug:
https://bugzilla.mozilla.org/show_bug.cgi?id=829896

The example given is this (`print()` call added for convenience):
```javascript
Int32Array.prototype[17] = "ehrmagerd"
print((new Int32Array(8))[17]);
```

As described in the bug report, the correct behavior is to return undefined.  Duktape prints the assigned string instead.

**Comment by svaarala** [comment]
Good catch - I've written the code originally using the Khronos spec and out-of-bounds accesses are considered normal property accesses.

**Comment by fatcerberus** [comment]
Do note that this only affects indexing - it seems named property access is unaffected and is allowed to consult the prototype.  I imagine that won't be too fun to handle, given that *all* property lookups are canonically keyed by string...

Honestly the ES specs are all so full of inconsistencies like this that I'm glad you're the one writing the JS engine and not me. I'd be afraid to actually contribute because it would be way too easy to end up breaking the spec in some bizarre corner case... :)

I figured I should bring this one up though in particular because as noted in the Bugzilla thread, the current behavior breaks asm.js.

**Comment by svaarala** [comment]
Yeah, I'll give the ES6 spec a read-through with respect to the correct behavior and compare to other engines too. The current TypedArray stuff is not 100% compliant in other ways too, so there are probably other ES6 issues to fix also (coercions may not be 100% compliant etc). But it's good you reported this, thanks.

**Comment by svaarala** [comment]
There are a few other exotic index related issues too:

- For a `view = new Uint8Array(10)` the expression `10 in view` should be false even if `10` is inherited (http://www.ecma-international.org/ecma-262/6.0/#sec-integer-indexed-exotic-objects-hasproperty-p).
- Object.defineProperty() has specific behavior for buffer object array indices; current implementation doesn't handle that correctly (http://www.ecma-international.org/ecma-262/6.0/#sec-integer-indexed-exotic-objects-defineownproperty-p-desc).

There were already mentioned in the issue / link:

- Direct [[Get]] on an exotic index object should not traverse to parent (http://www.ecma-international.org/ecma-262/6.0/#sec-integer-indexed-exotic-objects-get-p-receiver). Same in practice if [[Get]] is on a value which inherits from an exotic index object.
- Direct [[Set]] on an exotic index object should not traverse to parent (http://www.ecma-international.org/ecma-262/6.0/#sec-integer-indexed-exotic-objects-set-p-v-receiver). Same in practice if [[Set]] is on a value which inherits from an exotic index object.

**Comment by svaarala** [comment]
Interesting, ES6 behavior for the index string `"-0"` is:

- The string is considered a string index (i.e. triggers the special typed array behavior).
- However, it's always considered out-of-bounds and won't be associated with `view[0]`.
- So, accesing `view['-0']` always returns undefined, and won't traverse to parent.

This won't be a trivial fix because code needs to handle that specially. I'll maybe fix that separately.

My old Node.js (v0.12.1) doesn't handle this correctly, but Chrome and Firefox do. Newer Node.js would thus quite likely handle this correctly too.

**Comment by svaarala** [comment]
Hmm, this requires a bit more changes than I thought: the special behavior affects `[[Get]]` which will terminate a prototype chain lookup when encountering an exotic index object. However, `[[GetOwnProperty]]` will still report those exotic properties to be undefined (missing). I'll chip away on this in the coming weeks (some more pressing stuff to do first) :)

**Comment by fatcerberus** [comment]
The special casing for negative zero seems odd.  I guess because the strings "0" and "-0" are different keys they added a special case so that negative zero isn't treated the same as zero when it wouldn't be for, e.g. normal arrays.

**Comment by svaarala** [comment]
Right, but to match array behavior `"-0"` would be a non-index key and simply inherit from the prototype:

```js
Array.prototype['-0'] = 'dummy';
var arr = [];
arr['-0'];  // -> 'dummy' on a relatively recent Chrome; undefined for typed arrays
```

**Comment by fatcerberus** [comment]
Good point.  Yeah, no idea what the Ecmascript committee was getting at with that.  Oh well. :P
